src/coreclr/jit/codegencommon.cpp

```diff
@@ -7318,6 +7318,55 @@ void CodeGen::genIPmappingAddToFront(IPmappingDscKind kind, const DebugInfo& di,
 #endif // DEBUG
 }
 
+//------------------------------------------------------------------------
+// genIPmappingUpdateForRemovedInstruction: Update the IP mapping table for a removed instruction.
+// If the last IP mapping corresponds to the location immediately after the removed instruction, then
+// update the mapping to the location before the removed instruction.
+//
+// Arguments:
+//    loc - the emitter location we compare against
+//    removedCodeSize - the size of the instruction being removed
+//
+void CodeGen::genIPmappingUpdateForRemovedInstruction(emitLocation loc, unsigned removedCodeSize)
```

Comment:
@BruceForstall : I don't think this function is doing the right thing.
"update the mapping to the location before the removed instruction." seems wrong....


```
Generating: N027 (  1,  1) [000002] -----------                    t2 =    LCL_VAR   byref  V00 arg0         u:1 x19 REG x19 $80
                                                                        /--*  t2     byref  
Generating: N029 (  5,  4) [000004] DA---------                         *  STORE_LCL_VAR byref  V05 loc3          NA REG NA
IN0009:                           str     x19, [fp, #0x68]	// [V05 loc3]
Added IP mapping: 0x0008 STACK_EMPTY (G_M51615_IG02,ins#9,ofs#36)
Generating: N031 (???,???) [000059] -----------                            IL_OFFSET void   INLRT @ 0x008[E-] REG NA
Generating: N033 (  1,  1) [000005] -----------                    t5 =    LCL_VAR   byref  V00 arg0         u:1 x19 (last use) REG x19 $80
                                                                        /--*  t5     byref  
Generating: N035 (  1,  3) [000047] DA---------                         *  STORE_LCL_VAR long   V12 tmp4         d:1 x19 REG x19
							V00 in reg x19 is becoming dead  [000005]
							Live regs: 380000 {x19 x20 x21} => 300000 {x20 x21}
							Live vars: {V00 V01 V10} => {V01 V10}
							Byref regs: 180000 {x19 x20} => 100000 {x20}
							V12 in reg x19 is becoming live  [000047]
							Live regs: 300000 {x20 x21} => 380000 {x19 x20 x21}
							Live vars: {V01 V10} => {V01 V10 V12}
Added IP mapping: 0x000B STACK_EMPTY (G_M51615_IG02,ins#9,ofs#36)
Generating: N037 (???,???) [000060] -----------                            IL_OFFSET void   INLRT @ 0x00B[E-] REG NA
Generating: N039 (  1,  1) [000009] -----------                    t9 =    LCL_VAR   byref  V01 arg1         u:1 x20 REG x20 $81
                                                                        /--*  t9     byref  
Generating: N041 (  5,  4) [000011] DA---------                         *  STORE_LCL_VAR byref  V07 loc5          NA REG NA
Removing saved instruction in G_M51615_IG02:
> IN0009:                           str     x19, [fp, #0x68]	// [V05 loc3]
Updating last IP mapping: 0x000B STACK_EMPTY (G_M51615_IG02,ins#9,ofs#36)
                      to: 0x000B STACK_EMPTY (G_M51615_IG02,ins#8,ofs#32)
IN000a:                           stp     x20, x19, [fp, #0x60]
```
In the above code, an `str` is created at position 9. This also creates an IP mapping with `(G_M51615_IG02,ins#9,ofs#36)`.
A few lines later, the `str` is deleted and replaced with `stp`.
The `genIPmappingUpdateForRemovedInstruction()` causes that mapping to be updated from instruction 9 to instruction **8**.
I think the IP mapping needs to stay where it is - maybe `genIPmappingUpdateForRemovedInstruction()` just needs deleting? When I tried this, it go rid of the assert issues for me.

As a second issue, there is a gap in the instructions. In the example above, the `stp` becomes instruction 10 and 9 stays unused. I _think_ that's ok?
```
IN0006: 000040  F90023A0          str     x0, [fp, #0x40]
IN0007: 000044  AA1D03E0          mov     x0, fp
IN0008: 000048  F9002BA0          str     x0, [fp, #0x50]
IN000a: 00004C  A9064FB4          stp     x20, x19, [fp, #0x60]
IN000b: 000050  2A1F03E0          mov     w0, wzr
```

Reply:
> I think the IP mapping needs to stay where it is - maybe genIPmappingUpdateForRemovedInstruction() just needs deleting?

I think you're right: if the mapping was introduced to represent the `str` instruction (in this example), then it doesn't make sense for the mapping to be updated to represent the previous instruction.

However, it's possible that the `stp` instruction (the "replacement") should create an appropriate mapping. Not including the mapping is probably fine (maybe it creates a degraded optimized debugging experience... which isn't great anyway).

The INxxxx value (idNum) is debug only and only used for JitDump output, and in one case in an assert that requires it to be in increasing order in the code stream, so the "missing" number is fine. Actually, it might be good, since it indicates (subtly...) something was removed.

Reply:
> However, it's possible that the `stp` instruction (the "replacement") should create an appropriate mapping. Not including the mapping is probably fine (maybe it creates a degraded optimized debugging experience... which isn't great anyway).

After creating the stp it adds an IP mapping for that instruction too (it was missing from the dump above, so I've updated that comment).
So, by skipping `genIPmappingUpdateForRemovedInstruction()` you end up with:

```
IN000a:                           stp     x20, x19, [fp, #0x60]
IP mapping: 0x0008 STACK_EMPTY (G_M51615_IG02,ins#9,ofs#36)
IP mapping: 0x000E STACK_EMPTY (G_M51615_IG02,ins#9,ofs#36)
```

I'm not sure how the IP mappings are used, but that feels vaguely sensible.

Reply:
The problem that led me to introduce `genIPmappingUpdateForRemovedInstruction()` is that sometimes we remove an instruction (`str`), then when we add a new one (`stp`) the instrDesc is larger, doesn't fit in the currently used IG, so we create a new IG, at which point the previously captured `emitLocation` pointing in the old IG is no longer valid. In most cases where the replacement instruction fits in the existing IG, it kind of makes sense to leave both mappings. But the semantics of "remove last instruction" should, IMO, remove the now incorrect mapping, and the newly generated replacement instruction should generate whatever mappings it requires.